proposals/NNNN-shader-execution-reordering.md

```diff
@@ -0,0 +1,1388 @@
+over the reordering itself. These characteristics make `ReorderThread` a
+versatile tool for improving performance in a variety of workloads that suffer
+from divergent execution or data access.
+
+Similar to `TraceRay` and `CallShader`, the behavior of the actual reorder
+operation at any reorder point is implementation specific. An implementation
+may, for instance, not reorder at all, reorder within some local domain
+(threads on a local processor), reorder only within a certain time window,
+reorder globally across the entire GPU and the entire dispatch, or any
+variation thereof. The general target is for implementations to make a _best
+effort_ to extract as much coherence as possible, while keeping the overhead
+of reordering low enough to be practical for use in typical raytracing
+scenarios.
+
+---
+
+### Divergence behavior
+
+An implementation may only reorder threads whose control flow arrives at a
+reorder point. Threads that do not arrive at a reorder point are guaranteed to
+not migrate, even if neighboring threads in the same wave do migrate.
+
+From the physical wave's point of view, an implementation may remove, replace,
+or do nothing with threads that arrive at a reorder point. It is also legal
+for an implementation to replace terminated threads in a wave if any of its
+threads arrive at a reorder point. This may indirectly affect threads that
+conditionally did not invoke a reorder point, as illustrated in the following
+code snippet:
+
+```C++
+int MyFunc(int coherenceCoord)
+{ 
+    int A = WaveActiveBallot(true); 
+    if (WaveIsFirstLane()) 
+        ReorderThread(coherenceCoord, 32); 
+    int B = WaveActiveBallot(true); 
+    return A - B; 
+}
+```
+
+In this example, a number of different things could happen:
+- If the implementation does not honor `ReorderThread` at all, the function
+will most likely return zero, as the set of threads before and after the
+conditional reorder would be the same. 
+- If the implementation reorders threads invoking `ReorderThread` but does not
+replace them, B will likely be less than A for threads not invoking
+`ReorderThread`, while the reordered threads will likely resume execution with
+a newly formed full wave, thereby obtaining `A <= B`.
+- If the implementation replaces threads in a wave, the threads not
+participating in the reorder may possibly be joined by more threads than were
+removed from the wave, again observing `A <= B`.
+
+---
+
+### Memory coherence and visibility
+
+Due to the existence of non-coherent caches on most modern GPUs, an
+application must take special care when communicating information across
+reorder points through memory (UAVs).
+
+Specifically, if UAV stores or atomics are performed on one side of a reorder
+point, and on the other side the data is read via non-atomic UAV reads, the
+following is required:
+- The UAV must be declared `[globallycoherent]`.
+- The UAV writer must issue a `DeviceMemoryBarrier` between the write and the
+reorder point.
```

Comment:
Requiring global memory coherence can be a large hit to performance. Especially since a driver can decide not to sort if the workload is coherent (or sort in a more local manner).
Using `globallycoherent` and `DeviceMemoryBarrier` makes it impossible for a driver to omit the synchronization work when it’s not needed.

I propose that we add a new UAV attribute and barrier scope that relates the synchronization to reordering. So,
- The UAV must be declared `[reordercoherent]`
- The UAV writer must issue a `Barrier(UAV_MEMORY, THREAD_REORDER_SCOPE)` between the write and the
reorder point.

That way, when a driver decides not to reorder, the barrier can be a no-op.

Reply:
This seems useful at first glance, but expanding the memory model may overly broaden the scope of this proposal. It might be better suited as a separate proposal.

It is also unclear what use case [reordercoherent] is addressing. A UAV marked as [reordercoherent] can only communicate within the same DispatchRaysIndex, across any reordering. This essentially makes it a local scratch pad, which could be more efficiently handled using live state in the shader program (or as payload values if communicated with other shaders).

Reply:
> This seems useful at first glance, but expanding the memory model may overly broaden the scope of this proposal. It might be better suited as a separate proposal.

It needs one more enum value in [`SEMANTIC_FLAG`](https://microsoft.github.io/hlsl-specs/proposals/0018-work-graphs.html#barrier) and (optionally? see below) a `[reordercoherent]` marker on UAVs.

Vulkan also has a separate scope `ShaderCallKHR` for this.

Requiring global coherence is surely fine for some implementations but may incur a performance hit in other implementations.

> It is also unclear what use case [reordercoherent] is addressing. A UAV marked as [reordercoherent] can only communicate within the same DispatchRaysIndex, across any reordering.

The marker was meant to connect the barrier to certain UAVs. This allows UAVs that are not explicitly marked to skip synchronization.
I suppose an alternative is to not require marking UAVs but expect all bound UAVs to be synchronized with a reorder-scope barrier.
Although this may be expensive if reordering happens globally. Then this would be equivalent to mark all UAVs as `[globallycoherent]`.

Reply:
> The marker was meant to connect the barrier to certain UAVs. This allows UAVs that are not explicitly marked to skip synchronization.

This part I understand. What I would like to understand is the use case for a reorder-scope barrier. Since it can only be used to synchronize within the same DispatchRaysIndex, the data that is written/read to/from the reorder-scope coherent UAV can instead be stored as live state or in payload during intermediate steps and then be written once at the end (if needed). This seems more efficient overall.

Reply:
From last week's meeting, AMD proposal for a Reorder Coherence domain: just add new enums for UAV attribute, and barrier ( ) enum param.

@rasmusnv @Flakebi can we confirm the decision here?

Reply:
Yes. "Memory coherence and visibility" has been updated with the proposal.

Reply:
Yes, thanks for that.
The current text reads fine, I think it would be good to add that the new `REORDER_SCOPE` value is added to the `SEMANTIC_FLAG` enum and that `[reordercoherent]` is a new storage class that can be annotated on buffers.
Just to be clear to readers that these two are added in this proposal.